### ISSUE TITLE
Weights and positive weights for labels in Flux losses

### ISSUE THREAD

**Issue by gabrielpreviato** [FluxML/Flux.jl#2153]
### Motivation and description

Hello folks, I searched in Flux issues and documentations but couldn't find anything about having (or a reason for not having) weights for labels in the native losses functions.

Is it too much boilerplate for the Flux losses? Looking at their code they are generally very clean and concise.

If the recommendation is to build your own loss in case you need weighted labels, shouldn't we have this maybe more explicitly documented with one or more examples?

### Possible Implementation

One can easily implement this with something like:

```julia
function binarycrossentropy(ŷ, y; agg = mean, ϵ = epseltype(ŷ), weights=1)
    _check_sizes(ŷ, y, weitghs) # for checking if the weights are also valid
    agg(@.(weights'*(-xlogy(y, ŷ + ϵ) - xlogy(1 - y, 1 - ŷ + ϵ))))
end
```

A simple example:
```
julia> x = [[0.8] [0.1]]
1×2 Matrix{Float64}:
 0.8  0.1

julia> y = [[1.0] [0.0]]
1×2 Matrix{Float64}:
 1.0  0.0

julia> binarycrossentropy(x, y)
0.16425203348601775

julia> binarycrossentropy(x, y, weights=[0.75, 1.25])
0.1495291540289698
```

**Comment by CarloLucibello** [comment]
You can do that by passing a custom aggregation function that includes the weights,it is briefly mentioned [in the docs](http://fluxml.ai/Flux.jl/stable/models/losses/#man-losses), something like
```julia
loss(ŷ, y, agg = x -> mean(weights .* x))    # weighted mean
```

**Comment by gabrielpreviato** [comment]
Yes, this would work for the weights in all labels, but if you want it for the positive labels it wouldn't work, right?

Something like:
```julia
function binarycrossentropy(ŷ, y; agg = mean, ϵ = epseltype(ŷ), pos_weights=1)
    _check_sizes(ŷ, y)
    # or _check_sizes(ŷ, y, weights) for checking if the weights are valid
    agg(@.(-pos_weights*xlogy(y, ŷ + ϵ) - xlogy(1 - y, 1 - ŷ + ϵ)))
end
```

**Comment by CarloLucibello** [comment]
for that case you can do
```julia
weights = @. y * pos_weight + (1 - y) * neg_weight
logitbinarycrossentropy(ŷ, y; agg = x -> mean(weights .* x))
```

**Comment by gabrielpreviato** [comment]
Nice one! I think it would be good to have it there. From my experience, coming from other Python frameworks they have it built-in, so I think showing this more "Fluxian" way of doing a nice addition.

**Comment by mcabbott** [comment]
We could also consider adding some way to add weights to any function, something like this?  

```julia
julia> function weight(loss::Function, yhat, y, weights::AbstractArray; agg=mean, kw...)
         raw = loss(yhat, y; agg=identity, kw...)
         length(weights) == size(yhat, ndims(yhat)) || error("bad size")
         row = reshape(weights, ntuple(_->1, ndims(yhat)-1)..., :)
         λ = length(weights) ./ sum(weights)
         agg(raw .* row .* λ)
       end
weight (generic function with 1 method)

julia> yhat = rand(3,4); y = rand(3,4); Flux.crossentropy(yhat, y)
1.07989374514704

julia> weight(Flux.crossentropy, yhat, y, ones(4))
1.07989374514704

julia> weight(Flux.crossentropy, yhat, y, [1 0 0 0])
0.6339709254399699

julia> Flux.crossentropy(yhat[:,1:1], y[:,1:1])
0.6339709254399699
```
(Or it could read `f = weight(Flux.crossentropy, [1,0,0,0])` then `f(yhat, y)`.)

**Comment by ToucheSir** [comment]
One issue with the design of a weights API is that there's ambiguity between per-sample and per-class weights. PyTorch solves this by...not solving it and having a `weight` param that means [different](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html) [things](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss) across different losses (if indeed they expose such a parameter at all).
